lib/cadet/assessments.ex

```diff
@@ -0,0 +1,590 @@
+      from(
+        s in Submission,
+        join: dg in Group,
+        on: s.student_id == dg.student_id,
+        where: s.status == "submitted" and dg.staff_id == ^staff.id,
+        preload: [student: [:user], mission: []]
+      )
+    )
+  end
+
+  def get_mission_and_questions(id) do
+    Repo.one(
+      from(
+        a in Mission,
+        where: a.id == ^id,
+        left_join: q in Question,
+        on: q.mission_id == ^id,
+        preload: [:question]
+      )
+    )
+  end
+
+  def get_mission(id) do
+    Repo.get(Mission, id)
+  end
+
+  def create_question(params, type, mission_id) do
+    mission = get_mission(mission_id)
+    create_question_for_mission(params, type, mission)
+  end
+
+  def create_question_for_mission(params, type, mission) do
+    Repo.transaction(fn ->
+      mission = Repo.preload(mission, :questions)
+      questions = mission.questions
+
+      changeset =
+        params
+        |> Map.merge(%{type: type})
+        |> build_question()
+        |> put_assoc(:mission, mission)
+        |> put_display_order(questions)
+
+      case Repo.insert(changeset) do
+        {:ok, question} -> question
+        {:error, changeset} -> Repo.rollback(changeset)
+      end
+    end)
+  end
+
+  def find_submission(mission = %mission{}, student) do
+    find_submission(mission.id, student)
+  end
+
+  def find_submission(mission_id, student) do
+    Repo.one(
+      from(
+        submission in Submission,
+        where: submission.mission_id == ^mission_id and submission.student_id == ^student.id
+      )
+    )
+  end
+
+  def attempt_mission(id, student) when is_binary(id) do
+    mission = Repo.get(Mission, id)
+
+    if mission == nil do
+      {:error, :mission_not_found}
+    else
+      attempt_mission(mission, student)
+    end
+  end
+
+  def attempt_mission(mission, student) do
+    # Check for existing submission by student
+    submission = find_submission(mission, student)
+
+    if submission != nil do
+      {:ok, submission}
+    else
+      create_submission(mission, student)
+    end
+  end
+
+  def can_open?(id, user, student) do
+    can_attempt?(id, user) && find_submission(id, student) != nil
+  end
+
+  def can_attempt_mission?(mission, user) do
+    if user.role == :staff do
+      true
+    else
+      mission.is_published && opened?(mission)
+    end
+  end
+
+  def can_attempt?(id, user) do
+    mission = get_mission(id)
+    can_attempt_mission?(mission, user)
+  end
+
+  def opened?(mission) do
+    timezone = Timezone.get("Asia/Singapore", Timex.now())
+    date = Timezone.convert(mission.open_at, timezone)
+    Timex.before?(date, Timex.now())
+  end
+
+  def prepare_workspace(id, question_order, student) when is_binary(id) do
+    mission = Repo.get(Mission, id)
+
+    if mission == nil do
+      {:error, :mission_not_found}
+    else
+      prepare_workspace(mission, question_order, student)
+    end
+  end
+
+  def prepare_workspace(mission, question_order, student) do
```

Comment:
Not sure if this belongs here. Might be more appropriate to be put in the controller. @evansb what do you think?